src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttServerHandler.java

```diff
@@ -91,15 +110,76 @@ private void handleConnectMessage(ChannelHandlerContext ctx) {
         ctx.writeAndFlush(connAckMessage);
     }
 
+    /**
+     * Send a MQTT PUBACK message to the client.
+     *
+     * @param ctx ChannelHandlerContext instance
+     * @param packetId packet identifier
+     */
+    private void sendPubAckMessage(ChannelHandlerContext ctx, int packetId) {
+        MqttMessage pubAckMessage = MqttMessageBuilders.pubAck()
+                .packetId(packetId)
+                .build();
+
+        ctx.writeAndFlush(pubAckMessage);
+    }
+
     /**
      * Handle the case when a client sent a MQTT PUBLISH message type.
      *
      * @param ctx            ChannelHandlerContext instance
      * @param publishMessage represents a MqttPublishMessage
      */
     private void handlePublishMessage(ChannelHandlerContext ctx, MqttPublishMessage publishMessage) {
-        logger.info("MQTT Topic: {}", publishMessage.variableHeader().topicName());
-        logger.info("Kafka Mapped Topic: {}", mqttKafkaMapper.map(publishMessage.variableHeader().topicName()));
-        logger.info("Message: {}", publishMessage.payload().toString(Charset.defaultCharset()));
+        // get QoS level from the MqttPublishMessage
+        MqttQoS qos = MqttQoS.valueOf(publishMessage.fixedHeader().qosLevel().value());
+
+        // perform topic mapping
+        String mappedTopic = mqttKafkaMapper.map(publishMessage.variableHeader().topicName());
+
+        // build the Kafka record
+        ProducerRecord<String, ByteBuf> record = new ProducerRecord<>(mappedTopic,
+                publishMessage.payload());
+
+        // get the appropriate Kafka producer according to the QoS level
+        BridgeKafkaProducer producer = this.getProducer(qos);
+
+        // send the record to the Kafka topic
+        switch (qos) {
+            case AT_MOST_ONCE -> {
+                producer.send(record);
+                logger.info("Message sent to Kafka on topic {}", record.topic());
+            }
+            case AT_LEAST_ONCE -> {
+                CompletionStage<RecordMetadata> result = producer.send(record);
+                // wait for the result of the send operation
+                result.whenComplete((metadata, error) -> {
+                    if (error != null) {
+                        logger.error("Error sending message to Kafka: ", error);
+                    } else {
+                        logger.info("Message sent to Kafka on topic {} with offset {}", metadata.topic(), metadata.offset());
+                        // send PUBACK message to the client
+                        sendPubAckMessage(ctx, publishMessage.variableHeader().packetId());
+                    }
+                });
+            }
+            case EXACTLY_ONCE -> logger.warn("QoS level EXACTLY_ONCE is not supported yet");
```

Comment:
```suggestion
            case EXACTLY_ONCE -> logger.warn("QoS level EXACTLY_ONCE is not supported");
```

Let's not provide future expectation :-)